File: genes.py
import math
from numpy.core.fromnumeric import sort
from scipy.special import expit as sigmoid
import numpy as np
import util
import copy
import json


class RandomlyTrue:
    def __bool__(self):
        return util.flipCoin(0.5)
    instance = None


RandomlyTrue.instance = RandomlyTrue()

def random_uniform0(half_range):
    return np.random.normal(0, half_range)
    # return np.random.uniform(-half_range, half_range)

class Genes:

    class Metaparameters:
        def __init__(self,
                     c1=1, c2=1, c3=1,
                     perturbation_chance=0.8,
                     perturbation_stdev=0.1,
                     reset_weight_chance=0.1,
                     new_link_chance=0.3,
                     bias_link_chance=0.01,
                     new_link_weight_stdev=1,
                     new_node_chance=0.03,
                     disable_mutation_chance=0.1,
                     enable_mutation_chance=0.25):
            self.innovation_number = 0

            def none_or(value, default_value):
                return default_value if value is None else value
            self.c1 = none_or(c1, 0.1)
            self.c2 = none_or(c2, 0.1)
            self.c3 = none_or(c3, 0.1)
            self.new_link_chance = none_or(new_link_chance, 0.1)
            self.bias_link_chance = none_or(bias_link_chance, 0.1)
            self.new_link_weight_stdev = none_or(new_link_weight_stdev, 1)
            self.new_node_chance = none_or(new_node_chance, 0.1)
            self.perturbation_chance = none_or(perturbation_chance, 0.1)
            self.reset_weight_chance = none_or(reset_weight_chance, 0.5)
            self.perturbation_stdev = none_or(perturbation_stdev, 0.1)
            self.disable_mutation_chance = none_or(disable_mutation_chance, 0.1)
            self.enable_mutation_chance = none_or(enable_mutation_chance, 0.1)
            self._connections = {}
            self._node_splits = {}

        def _increment_innovation(self):
            self.innovation_number += 1
            return self.innovation_number

        def reset_tracking(self):
            self._connections = {}
            self._node_splits = {}

        def register_connection(self, in_node, out_node):
            pair = (in_node, out_node)
            innovation_number = self._connections.get(pair, None)
            if innovation_number is None:
                innovation_number = self._increment_innovation()
                self._connections[pair] = innovation_number
            return innovation_number

        def register_node_split(self, in_node, out_node, between_node):
            tuple = (in_node, out_node, between_node)
            innovation_numbers = self._node_splits.get(tuple, None)
            if innovation_numbers is None:
                leading = self._increment_innovation()
                trailing = self._increment_innovation()
                innovation_numbers = (leading, trailing)
                self._node_splits[tuple] = innovation_numbers
            return innovation_numbers

        def load(in_stream, decoder=json):
            as_json = decoder.load(in_stream)
            ret = Genes.Metaparameters(
                c1=as_json.get("c1"),
                c2=as_json.get("c2"),
                c3=as_json.get("c3"),
                new_link_weight_stdev=as_json.get("new_link_weight_stdev"),
                new_node_chance=as_json.get("new_node_chance"),
                perturbation_chance=as_json.get("perturbation_chance"),
                perturbation_stdev=as_json.get("perturbation_stdev"),
                reset_weight_chance=as_json.get("reset_weight_chance"),
                disable_mutation_chance=as_json.get("disable_mutation_chance"),
                enable_mutation_chance=as_json.get("enable_mutation_chance")
            )
            if "innovation_number" in as_json:
                ret.innovation_number = as_json["innovation_number"]
            return ret

        def save(self, out_stream, encoder=json):
            out = {
                "innovation_number": self.innovation_number,
                "c1": self.c1,
                "c2": self.c2,
                "c3": self.c3,
                "new_link_chance": self.new_link_chance,
                "bias_link_chance": self.bias_link_chance,
                "new_link_weight_stdev": self.new_link_weight_stdev,
                "new_node_chance": self.new_node_chance,
                "perturbation_chance": self.perturbation_chance,
                "perturbation_stdev": self.perturbation_stdev,
                "reset_weight_chance": self.reset_weight_chance,
                "disable_mutation_chance": self.disable_mutation_chance,
                "enable_mutation_chance": self.enable_mutation_chance,
            }
            out_stream.write(encoder.dumps(out))
            out_stream.flush()

    _IN_NODE = 0
    _OUT_NODE = 1
    _WEIGHT = 2
    _ENABLED = 3
    _INNOV_NUMBER = 4
    BIAS_INDEX = 0

    def __init__(self, num_sensors_or_copy, num_outputs=None, metaparameters=None):
        if isinstance(num_sensors_or_copy, Genes):
            to_copy = num_sensors_or_copy
            self._num_sensors = to_copy._num_sensors
            self._num_outputs = to_copy._num_outputs
            self._dynamic_nodes = copy.deepcopy(to_copy._dynamic_nodes)
            self._connections = copy.deepcopy(to_copy._connections)
            self._metaparameters = to_copy._metaparameters
            self._connections_sorted = to_copy._connections_sorted
        else:
            self._num_sensors = num_sensors_or_copy
            self._num_outputs = num_outputs
            self._dynamic_nodes = []
            for _ in range(num_outputs):
                self._dynamic_nodes.append([])
            self._connections = []
            self._metaparameters = metaparameters
            self._connections_sorted = True
        self.fitness = 0

    def feed_sensor_values(self, values, neurons=None):
        """ Run the network with the given input through the given neurons (creates them if not given), returns neuron values """
        if neurons is None:
            neurons = [0] * (len(self._dynamic_nodes) + self._num_sensors + 1)
        assert len(values) == self._num_sensors, "invalid number of inputs"
        neurons[0] = 1  # BIAS node
        for i in range(self._num_sensors):
            neurons[i + 1] = values[i]

        def feed(node_index):
            node = self._dynamic_nodes[node_index]
            neuron_index = node_index + self._num_sensors + 1
            has_connections = False
            sum = 0

            for connection_index in node:
                in_node, out_node, weight, enabled, innov = self._connections[connection_index]
                # assert out_node = neuron_index
                if enabled:
                    has_connections = True
                    sum += neurons[in_node] * weight
            if has_connections:
                neurons[neuron_index] = sigmoid(sum)

        for hidden_node_index in range(self._num_outputs, len(self._dynamic_nodes)):
            feed(hidden_node_index)
        for output_node_index in range(self._num_outputs):
            feed(output_node_index)

        return neurons

    def extract_output_values(self, neuron_values):
        """ Extracts the output values from the result of feed_sensor_values """
        return neuron_values[self._num_sensors + 1:self._num_sensors + self._num_outputs + 1]

    def _node_by_index(self, index):
        return self._dynamic_nodes[index - self._num_sensors - 1]

    def total_nodes(self):
        return 1 + self._num_sensors + len(self._dynamic_nodes)

    def input_node_index(self, input_index):
        return 1 + input_index

    def output_node_index(self, index):
        return 1 + self._num_sensors + index    

    def add_connection(self, input_index, output_index):
        incoming = self._node_by_index(output_index)
        for connection_index in incoming:
            connection = self._connections[connection_index]
            if connection[Genes._IN_NODE] == input_index:
                return self

        innovation_number = self._metaparameters.register_connection(input_index, output_index)
        connection = [input_index, output_index, random_uniform0(self._metaparameters.new_link_weight_stdev), True, innovation_number]
        incoming.append(len(self._connections))
        if len(self._connections) > 0 and innovation_number < self._connections[-1][Genes._INNOV_NUMBER]:
            self._connections_sorted = False
        self._connections.append(connection)
        return self

    def _add_connection(self):
        total_nodes = self.total_nodes()
        input_index = 0 if util.flipCoin(self._metaparameters.bias_link_chance) else util.random.randint(0, total_nodes - 1)
        output_index = util.random.randint(self._num_sensors, total_nodes - 1)
        self.add_connection(input_index, output_index)

    def _add_node(self):
        if len(self._connections) == 0:
            return
        connection = util.random.choice(self._connections)
        connection[Genes._ENABLED] = False
        in_node, out_node, _a, _b, _c = connection
        new_node = []
        self._dynamic_nodes.append(new_node)
        leading_innov, trailing_innov = self._metaparameters.register_node_split(in_node, out_node, self.total_nodes() - 1)
        leading = [in_node, self.total_nodes() - 1, 1, True, leading_innov]
        trailing = [self.total_nodes() - 1, out_node, connection[Genes._WEIGHT], True, trailing_innov]
        if len(self._connections) > 0 and leading_innov < self._connections[-1][Genes._INNOV_NUMBER]:
            self._connections_sorted = False
        self._connections.append(leading)
        self._connections.append(trailing)
        new_node.append(len(self._connections) - 2)
        self._node_by_index(out_node).append(len(self._connections) - 1)

    def _perturb(self):
        reset_chance = self._metaparameters.reset_weight_chance
        for connection in self._connections:
            if util.flipCoin(reset_chance):
                connection[Genes._WEIGHT] = random_uniform0(self._metaparameters.new_link_weight_stdev)
            else:
                connection[Genes._WEIGHT] += random_uniform0(self._metaparameters.perturbation_stdev)

    def _enable_mutation(self, enable):
        if len(self._connections) == 0:
            return
        connection = util.random.choice(self._connections)
        connection[Genes._ENABLED] = enable

    def mutate(self):
        """ Mutate the genes in this genome, returns self """
        if util.flipCoin(self._metaparameters.new_link_chance):
            self._add_connection()
        if util.flipCoin(self._metaparameters.new_node_chance):
            self._add_node()
        if util.flipCoin(self._metaparameters.perturbation_chance):
            self._perturb()
        # if util.flipCoin(self._metaparameters.disable_mutation_chance):
        #     self._enable_mutation(False)
        # if util.flipCoin(self._metaparameters.enable_mutation_chance):
        #     self._enable_mutation(True)
        return self

    def _sorted_connections(self):
        if self._connections_sorted:
            return self._connections
        return sorted(self._connections, key=lambda c: c[Genes._INNOV_NUMBER])

    def breed(self, other, self_more_fit=RandomlyTrue.instance):
        """ Creates a child from the result of breeding self and other genes, returns new child """
        ret = Genes(self._num_sensors, self._num_outputs, self._metaparameters)

        sconnections = self._sorted_connections()
        oconnections = other._sorted_connections()
        slen = len(sconnections)
        olen = len(oconnections)
        i = 0
        j = 0

        def turn_on_maybe(connection):
            if not connection[Genes._ENABLED] and util.flipCoin(self._metaparameters.enable_mutation_chance):
                connection[Genes._ENABLED] = True

        while i < slen and j < olen:
            sc = sconnections[i]
            so = oconnections[j]
            sci = sc[Genes._INNOV_NUMBER]
            soi = so[Genes._INNOV_NUMBER]
            if sci == soi:
                ret._connections.append(copy.copy(sc if util.flipCoin(0.5) else so))
                turn_on_maybe(ret._connections[-1])
                i += 1
                j += 1
            elif sci < soi:
                i += 1
                if self_more_fit:
                    ret._connections.append(copy.copy(sc))
                    turn_on_maybe(ret._connections[-1])
            else:
                j += 1
                if not self_more_fit:
                    ret._connections.append(copy.copy(so))
                    turn_on_maybe(ret._connections[-1])
        while i < slen:
            if self_more_fit:
                ret._connections.append(copy.copy(sconnections[i]))
                turn_on_maybe(ret._connections[-1])
            i += 1
        while j < olen:
            if not self_more_fit:
                ret._connections.append(copy.copy(oconnections[j]))
                turn_on_maybe(ret._connections[-1])
            j += 1
        max_node = 0
        for connection in ret._connections:
            max_node = max(max_node, connection[Genes._IN_NODE], connection[Genes._OUT_NODE])
        i = ret.total_nodes()
        while i <= max_node:
            ret._dynamic_nodes.append([])
            i += 1
        for index, connection in enumerate(ret._connections):
            ret._node_by_index(connection[Genes._OUT_NODE]).append(index)
        return ret

    def distance(self, other):
        assert self._metaparameters is other._metaparameters
        c1 = self._metaparameters.c1
        c2 = self._metaparameters.c2
        c3 = self._metaparameters.c3
        sconnections = self._sorted_connections()
        oconnections = other._sorted_connections()
        slen = len(sconnections)
        olen = len(oconnections)
        n = max(olen, slen)
        if n == 0:
            return 0
        disjoint = 0
        weight_difference = 0
        shared = 0
        i = 0
        j = 0
        while i < slen and j < olen:
            sc = sconnections[i]
            so = oconnections[j]
            sci = sc[Genes._INNOV_NUMBER]
            soi = so[Genes._INNOV_NUMBER]
            if sci == soi:
                weight_difference += abs(sc[Genes._WEIGHT] - so[Genes._WEIGHT])
                shared += 1
                i += 1
                j += 1
            else:
                disjoint += 1
                if sci < soi:
                    i += 1
                else:
                    j += 1
        excess = olen - j + slen - i
        return (c1 * excess / n) + (c2 * disjoint / n) + (0 if shared == 0 else c3 * weight_difference / shared)

    def clone(self):
        return Genes(self)

    def as_json(self):
        """ returns self as a dict """
        return {"nodeCount": self.total_nodes(), "inputCount": self._num_sensors, "outputCount": self._num_outputs, "connections": self._connections}

    def save(self, out_stream, encoder=json):
        """ save to the stream using the given encoder, encoder must define dumps function that takes in a JSON-like object"""
        as_json = self.as_json()
        out_stream.write(encoder.dumps(as_json))
        out_stream.flush()

    def load_from_json(json_object, metaparameters):
        """ loads from a dict-like object """
        ret = Genes(json_object["inputCount"], json_object["outputCount"], metaparameters)
        to_add = json_object["nodeCount"] - ret.total_nodes()
        for _ in range(to_add):
            ret._dynamic_nodes.append([])
        connections = json_object["connections"]
        count = 0
        ret._connections = connections
        connections.sort(key=lambda c: c[Genes._INNOV_NUMBER])
        for in_node, out_node, weight, enabled, innov in connections:
            ret._node_by_index(out_node).append(count)
            count += 1
        return ret

    def load(in_stream, metaparameters, decoder=json):
        """ load from stream using given decoder, decoder must define load function that takes in a stream and returns a dict-like object"""
        as_json = decoder.load(in_stream)
        return Genes.load_from_json(as_json, metaparameters)

    def setFitness(self, fitness):
        self.fitness = fitness

    def getFitness(self):
        return self.fitness

File: Genes_.cpp
// Genes.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <vector>
#include <random>
#include "Genes_.h"
#include <iostream>

auto& random_generator() noexcept
{
	thread_local std::random_device rd;
	thread_local std::mt19937 gen(rd());
	return gen;
}

float random_uniform0(float range) noexcept
{
	std::uniform_real_distribution<float> dis(-range, range);
	return dis(random_generator());
}

bool flip_coin(float p) noexcept
{
	std::uniform_real_distribution<float> dis(0, 1);
	return dis(random_generator()) < p;
}

template<typename T, typename U>
auto randint(T min, U max) noexcept
{
	std::uniform_int_distribution<std::common_type_t<T, U>> dis(min, max);
	return dis(random_generator());
}

template<typename Range>
auto random_choice(Range& range) -> decltype(range.size(), range[0])
{
	return range[randint(0, range.size() - 1)];
}

void _Metaparameters::reset_tracking()
{
	_connections.clear();
	_node_splits.clear();
}

unsigned int _Metaparameters::register_connection(unsigned int in_node, unsigned int out_node)
{
	auto const pair = decltype(_connections)::key_type{ in_node, out_node };
	auto const [loc, inserted] = _connections.insert({ pair, 0 });
	if (inserted)
	{
		loc->second = ++innovation_number;
	}
	return loc->second;
}

std::pair<unsigned int, unsigned int> _Metaparameters::register_node_split(unsigned int in_node, unsigned int out_node, unsigned int between_node)
{
	auto const pair = decltype(_node_splits)::key_type{ in_node, {out_node, between_node} };
	auto const [loc, inserted] = _node_splits.insert({ pair, 0 });
	if (inserted)
	{
		loc->second = ++innovation_number;
		++innovation_number;
	}
	return { loc->second, loc->second + 1 };
}

void _Network::feed_sensor_values(float const* input, float* const neurons) const noexcept{
	neurons[0] = 1;
	std::copy(input, input + _num_inputs, neurons + 1);
	feed_sensor_values(neurons);
}

void _Network::feed_sensor_values(float* const neurons) const noexcept
{
	auto const feed = [&, dynamic_neurons = neurons + _num_inputs + 1](unsigned int node_index)
	{
		auto const& node = _dynamic_nodes[node_index];
		double sum = 0;
		bool has_connections = false;
		for (auto const connection_index : node)
		{
			auto const& connection = _connections[connection_index];
			if (connection.enabled)
			{
				has_connections = true;
				sum += neurons[connection.in_node] * double(connection.weight);
			}
		}
		if (has_connections)
		{
			dynamic_neurons[node_index] = std::tanh(sum);
		}
	};

	for (unsigned int i = _num_outputs; i < _dynamic_nodes.size(); ++i)
	{
		feed(i);
	}
	for (unsigned int i = 0; i < _num_outputs; ++i)
	{
		feed(i);
	}
}

_Network::_Network(unsigned int num_inputs, unsigned int num_outputs):_num_inputs(num_inputs), _num_outputs(num_outputs), _dynamic_nodes(num_outputs), _connections_sorted(true)
{}

unsigned int _Network::input_node_index(unsigned int input_index) const noexcept
{
	return input_index + 1;
}

unsigned int _Network::output_node_index(unsigned int output_index) const noexcept
{
	return output_index + input_count() + 1;
}

_Network::changed_t _Network::add_connection(unsigned int in_node_index, unsigned out_node_index, _Metaparameters& _metaparameters)
{
	if (!_metaparameters.allow_recurrent)
	{
		if (in_node_index == out_node_index)
		{
			return false;
		}
		if (_is_output_node_index(in_node_index))
		{
			if (out_node_index < in_node_index || _is_hidden_node_index(out_node_index))
			{
				std::swap(in_node_index, out_node_index);
			}
		}
		else if (_is_hidden_node_index(out_node_index) && out_node_index < in_node_index)
		{
			std::swap(in_node_index, out_node_index);
		}
	}
	auto& incoming = _dynamic_nodes[out_node_index - input_count() - 1];
	for (auto const& connection_index : incoming)
	{
		if (_connections[connection_index].in_node == in_node_index)
		{
			return false;
		}
	}
	auto const innovation_number = _metaparameters.register_connection(in_node_index, out_node_index);
	incoming.push_back(_connections.size());
	if (_connections.size() && innovation_number < _connections.back().innov_number)
	{
		_connections_sorted = false;
	}
	_connections.push_back({ in_node_index, out_node_index, random_uniform0(_metaparameters.new_link_weight_stdev), true, innovation_number });
	return true;
}

void _Network::perturb(_Metaparameters& metaparameters)
{
	auto const reset_chance = metaparameters.reset_weight_chance;
	auto const reset_std = metaparameters.new_link_weight_stdev;
	auto const pert_std = metaparameters.perturbation_stdev;
	for (auto& connection : _connections)
	{
		if (flip_coin(reset_chance))
		{
			connection.weight = random_uniform0(reset_std);
		}
		else
		{
			connection.weight += random_uniform0(pert_std);
		}
	}
}

bool _Network::mutate(_Metaparameters& metaparameters)
{
	bool changed = false;
	for (unsigned int i = 0; i < metaparameters.mutate_loop; ++i)
	{
		if (flip_coin(metaparameters.new_link_chance))
		{
			changed |= _add_connection(metaparameters);
		}
		if (flip_coin(metaparameters.new_node_chance))
		{
			changed |= _add_node(metaparameters);
		}
		if (flip_coin(metaparameters.perturbation_chance))
		{
			perturb(metaparameters);
			changed = true;
		}
		if (flip_coin(metaparameters.disable_mutation_chance))
		{
			_enable_mutation(false);
			changed = true;
		}
	}
	return changed;
}

_Network breed(_Network const& a, _Network const& b, double fitness_a, double fitness_b, _Metaparameters const& metaparameters)
{
	auto ret = _Network(a._num_inputs, a._num_outputs);
	std::vector<connection_t> _a_connections, _b_connections;
	auto const& a_connections = a._get_sorted_connections(_a_connections);
	auto const& b_connections = b._get_sorted_connections(_b_connections);
	auto const a_len = a_connections.size();
	auto const b_len = b_connections.size();
	std::size_t i = 0;
	std::size_t j = 0;
	auto ret_add_connection = [&](connection_t new_connection)
	{
		if (flip_coin(metaparameters.enable_mutation_chance))
		{
			new_connection.enabled = true;
		}
		ret._connections.push_back(new_connection);
	};
	while (i < a_len && j < b_len)
	{
		auto& sc = a_connections[i];
		auto& so = b_connections[j];
		auto const sci = sc.innov_number;
		auto const soi = so.innov_number;
		if (sci == soi)
		{
			ret_add_connection(flip_coin(0.5) ? sc : so);
			++i;
			++j;
		}
		else if (sci < soi)
		{
			++i;
			if (fitness_a > fitness_b || ((fitness_a == fitness_b) && flip_coin(0.5)))
			{
				ret_add_connection(sc);
			}
		}
		else
		{
			++j;
			if (fitness_a < fitness_b || ((fitness_a == fitness_b) && flip_coin(0.5)))
			{
				ret_add_connection(so);
			}
		}
	}
	if (fitness_a > fitness_b)
	{
		for (; i < a_len; ++i)
		{
			ret_add_connection(a_connections[i]);
		}
	}
	else if (fitness_a == fitness_b)
	{
		for (; i < a_len; ++i)
		{
			if (flip_coin(0.5))
			{
				ret_add_connection(a_connections[i]);
			}
		}
		for (; j < b_len; ++j)
		{
			if (flip_coin(0.5))
			{
				ret_add_connection(b_connections[j]);
			}
		}
	}
	else
	{
		for (; j < b_len; ++j)
		{
			ret_add_connection(b_connections[j]);
		}
	}

	auto const max_node = std::max(a._dynamic_nodes.size(), b._dynamic_nodes.size());
	ret._dynamic_nodes.resize(max_node);
	auto const dynamic_node_base = ret._dynamic_nodes.data() - ret.input_count() - 1;
	for (std::size_t i = 0; i < ret._connections.size(); ++i)
	{
		dynamic_node_base[ret._connections[i].out_node].push_back(static_cast<unsigned int>(i));
	}
	return ret;
}

double distance(_Network const& a, _Network const& b, _Metaparameters const& metaparameters)
{
	std::vector<connection_t> _a_con, _b_con;
	auto const& a_connections = a._get_sorted_connections(_a_con);
	auto const& b_connections = b._get_sorted_connections(_b_con);
	auto const alen = a_connections.size();
	auto const blen = b_connections.size();
	auto const n = std::max(alen, blen);
	if (n == 0)
	{
		return 0;
	}
	auto const c1 = metaparameters.c1;
	auto const c2 = metaparameters.c2;
	auto const c3 = metaparameters.c3;
	double disjoint = 0;
	double weight_difference = 0;
	double shared = 0;
	std::size_t i = 0;
	std::size_t j = 0;
	while (i < alen && j < blen)
	{
		auto const& sc = a_connections[i];
		auto const& so = b_connections[j];
		auto const sci = sc.innov_number;
		auto const soi = so.innov_number;
		if (sci == soi)
		{
			weight_difference += std::abs(double(sc.weight) - so.weight);
			shared += 1;
			++i;
			++j;
		}
		else
		{
			disjoint += 1;
			if (sci < soi)
			{
				++i;
			}
			else
			{
				++j;
			}
		}
	}
	auto const excess = double(blen - j + alen - i);
	return (c1 * excess / n) + (c2 * disjoint / n) + (shared == 0 ? 0.0 : c3 * weight_difference / shared);
}

std::vector<connection_t> const& _Network::_get_sorted_connections(std::vector<connection_t>& space) const
{
	if (_connections_sorted)
	{
		return _connections;
	}
	space = _connections;
	_sort_connections(space);
	return space;
}

bool _Network::_is_output_node_index(unsigned int index) const noexcept
{
	return index > input_count() && index < (1 + input_count() + output_count());
}

bool _Network::_is_hidden_node_index(unsigned int index) const noexcept
{
	return index >= 1 + input_count() + output_count();
}

_Network::changed_t _Network::_add_connection(_Metaparameters& _metaparameters)
{
	auto const total_nodes = total_neurons();
	auto const in_node = flip_coin(_metaparameters.bias_link_chance) ? 0 : randint(0, total_nodes - 1);
	auto const out_node = randint(_num_inputs + 1, total_nodes - 1);
	return add_connection(in_node, out_node, _metaparameters);
}

_Network::changed_t _Network::_add_node(_Metaparameters& _metaparameters)
{
	auto& connections = _connections;
	if (connections.size() == 0)
	{
		return false;
	}
	auto& connection = [&]() -> auto&
	{
		if (_metaparameters.allow_recurrent)
		{
			return random_choice(connections);
		}
		std::vector<unsigned int> choices;
		for (unsigned int i = 0; i < output_count(); ++i)
		{
			auto& node = _dynamic_nodes[i];
			choices.insert(choices.end(), node.begin(), node.end());
		}
		return connections[random_choice(choices)];
	}();
	if (!connection.enabled)
	{
		return false;
	}
	connection.enabled = false;
	auto const new_node_index = static_cast<unsigned int>(total_neurons());
	auto& new_node = _dynamic_nodes.emplace_back();
	auto const [leading_innov, trailing_innov] = _metaparameters.register_node_split(connection.in_node, connection.out_node, new_node_index);
	if (_connections.size() && leading_innov < _connections.back().innov_number)
	{
		_connections_sorted = false;
	}
	_connections.push_back({ connection.in_node, new_node_index, 1.f, true, leading_innov });
	_connections.push_back({ new_node_index, connection.out_node, connection.weight, true, trailing_innov });
	return true;
}

void _Network::_enable_mutation(bool enable)
{
	if (_connections.size())
	{
		random_choice(_connections).enabled = enable;
	}
}

void _Network::_synchronize_connections()
{
	_sort_connections(_connections);
	auto const dynamic_node_base = _dynamic_nodes.data() - input_count() - 1;
	for (std::size_t i = 0; i < _connections.size(); ++i)
	{
		dynamic_node_base[_connections[i].out_node].push_back(static_cast<unsigned int>(i));
	}
}